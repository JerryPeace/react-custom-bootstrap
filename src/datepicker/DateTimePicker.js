import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import Datetime from 'react-datetime';
import moment from 'moment';
import TmIcon from '../icon/TmIcon';

class DateTimePicker extends Component {
    static propTypes = {
        // icons and format
        backLabel: PropTypes.node,
        columnOfMonth: PropTypes.number,
        dateFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
        entryLabel: PropTypes.node,
        fullMonthLabel: PropTypes.bool,
        inputLabel: PropTypes.node,
        locale: PropTypes.string,
        minTimeGap: PropTypes.number,
        timeFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
        timeGapUnit: PropTypes.string,
        timeToggleLabelDown: PropTypes.node,
        timeToggleLabelUp: PropTypes.node,
        weekDayDisplayMode: PropTypes.string,

        // displaying controll
        closeOnSelect: PropTypes.bool,
        closeOnTab: PropTypes.bool,
        disabled: PropTypes.bool,
        inputProps: PropTypes.object,
        open: PropTypes.bool,

        // methods
        isValidDate: PropTypes.func,
        renderDay: PropTypes.func,
        validator: PropTypes.func,
        onChange: PropTypes.func,
        onFocus: PropTypes.func,

        // values
        defaultValue: PropTypes.any,
        value: PropTypes.any,
    };

    static defaultProps = {
        // icons and format
        backLabel: <TmIcon name='calendar' />,
        columnOfMonth: 3,
        dateFormat: 'YYYY-MM-DD',
        entryLabel: <TmIcon name='clock' />,
        fullMonthLabel: true,
        minTimeGap: 1,
        timeFormat: 'HH:mm:ss',
        timeGapUnit: 's',
        timeToggleLabelDown: <TmIcon name='chevron-down' />,
        timeToggleLabelUp: <TmIcon name='chevron-up' />,
        viewMode: 'days',
        weekDayDisplayMode: 'short',

        // displaying controll
        closeOnSelect: true,
        closeOnTab: true,
        disabled: false,
        inputProps: {},

        // methods
        isValidDate: (currentDate, selectedDate) => true,
        renderDay: (props, currentDate, selectedDate) => (
            <td {...props}><span {...props}>{currentDate.date()}</span></td>
        ),
    };

    constructor(props) {
        super(props);

        this.state = {
            currValue: props.defaultValue
        };

        this.bindHandlers();
    }

    bindHandlers() {
        this.getValue = ::this.getValue;
        this.isValid = ::this.isValid;
        this.isValidDate = ::this.isValidDate;
        this.isValidUTC = ::this.isValidUTC;
        this.onChange = ::this.onChange;
    }

    getValue() {
        if (this.props.value) {
            return this.props.value;
        } else {
            return this.state.currValue;
        }
    }

    isValid() {
        const selectedDate = this.getValue();
        return this.isValidDate(selectedDate) && this.props.validator(selectedDate);
    }

    isValidDate(currentDate, selectedDate) {
        if (!moment.isMoment(currentDate)) {
            currentDate = moment(currentDate);
        }

        return this.isValidUTC(currentDate) && this.props.isValidDate(currentDate, selectedDate);
    }

    isValidUTC(momentObject) {
        if (momentObject && momentObject.unix) {
            // Time should be after 1 January 1970 00:00:00 UTC
            return momentObject.unix() >= 0;
        }
        return false;
    }

    onChange(selectedDate) {
        if (!this.props.value) {
            this.setState({ currValue: selectedDate });
        }
        this.props.onChange && this.props.onChange(selectedDate);
    }

    renderInputLable() {
        // According to Picker type, different DEFAULT inputLabel is assigned.
        const { dateFormat, inputLabel } = this.props;
        if (inputLabel === undefined) {
            return dateFormat ? <TmIcon name='calendar' /> : <TmIcon name='clock' />;
        }
        return inputLabel;
    }

    render() {
        const {
            backLabel,
            className,
            closeOnSelect,
            closeOnTab,
            columnOfMonth,
            dateFormat,
            disabled,
            entryLabel,
            fullMonthLabel,
            inputProps,
            locale,
            minTimeGap,
            open,
            renderDay,
            timeFormat,
            timeGapUnit,
            timeToggleLabelDown,
            timeToggleLabelUp,
            validator,
            weekDayDisplayMode,
            onFocus
        } = this.props;

        const isDatePicker = !timeFormat;
        const isTimePicker = !dateFormat;
        const defaultClassName = className ? {} : {
            'rcb-date-picker': isDatePicker,
            'rcb-time-picker': isTimePicker
        };

        return (
            <Datetime
                ref='dateTime'
                backLabel={isTimePicker ? null : backLabel}
                className={classNames(className, defaultClassName)}
                closeOnSelect={closeOnSelect}
                closeOnTab={closeOnTab}
                columnOfMonth={columnOfMonth}
                dateFormat={dateFormat}
                disabled={disabled}
                entryLabel={entryLabel}
                fullMonthLabel={fullMonthLabel}
                inputLabel={this.renderInputLable()}
                inputProps={{...inputProps, disabled}}
                isValidDate={this.isValidDate}
                locale={locale}
                minTimeGap={minTimeGap}
                open={open}
                renderDay={renderDay}
                timeFormat={timeFormat}
                timeGapUnit={timeGapUnit}
                timeToggleLabelDown={timeToggleLabelDown}
                timeToggleLabelUp={timeToggleLabelUp}
                validator={validator}
                value={this.getValue()}
                weekDayDisplayMode={weekDayDisplayMode}
                onChange={this.onChange}
                onFocus={onFocus}
            />
        );
    }
}

export default DateTimePicker;
