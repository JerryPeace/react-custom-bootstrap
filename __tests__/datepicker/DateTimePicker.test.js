import React from 'react';
import sinon from 'sinon';
import moment from 'moment';
import {mount, shallow} from 'enzyme';
import DateTimePicker from 'datepicker/DateTimePicker';

describe('DateTimePicker', () => {
    const value = '2018-10-11 20:20:20';
    const newValue = '2018-11-12 10:10:10';
    const illegalValue = '2018-20-00 10:10:10';
    const isValidSpy = sinon.spy(DateTimePicker.prototype, 'isValid');
    const onChangeSpy = sinon.spy(DateTimePicker.prototype, 'onChange');

    test('should render correctly.', () => {
        const datepicker = shallow(
            <DateTimePicker value={value} />
        );

        expect(datepicker).toMatchSnapshot();
    });

    test('should be able to set default value.', () => {
        const datepicker = mount(<DateTimePicker defaultValue={value} />);
        const result = datepicker.node.getValue();
        expect(result).toEqual(value);
    });

    test('should be able to get date when invoking getValue method.', () => {
        const datepicker = mount(<DateTimePicker value={value} />);
        const result = datepicker.node.getValue();
        expect(result).toEqual(value);
    });

    test('should change value when invoking onChange method.', () => {
        const handleChange = (value) => {
            datepicker.setProps({ value: value });
        };

        const datepicker = mount(
            <DateTimePicker
                value={value}
                onChange={handleChange}
            />
        );

        expect(datepicker.prop('value')).toEqual(value);
        expect(onChangeSpy.callCount).toEqual(0);
        datepicker.node.onChange(newValue);
        expect(datepicker.prop('value')).toEqual(newValue);
        expect(onChangeSpy.callCount).toEqual(1);
    });

    test('should be able to validate the illegal date(string).', () => {
        const isValidDate = (currentDate) => {
            return moment(currentDate).isValid();
        };
        const datepicker = mount(
            <DateTimePicker
                value={illegalValue}
                isValidDate={isValidDate}
                validator={isValidDate}
            />
        );

        expect(isValidSpy.callCount).toEqual(0);
        expect(datepicker.node.isValid()).toEqual(false);
        expect(isValidSpy.callCount).toEqual(1);
    });

    test('should be able to validate the illegal date(moment object).', () => {
        const value = moment(illegalValue);
        const isValidDate = (currentDate) => {
            return moment(currentDate).isValid();
        };
        const datepicker = mount(
            <DateTimePicker
                value={illegalValue}
                isValidDate={isValidDate}
                validator={isValidDate}
            />
        );

        expect(isValidSpy.callCount).toEqual(1);
        expect(datepicker.node.isValid()).toEqual(false);
        expect(isValidSpy.callCount).toEqual(2);
    });
});
