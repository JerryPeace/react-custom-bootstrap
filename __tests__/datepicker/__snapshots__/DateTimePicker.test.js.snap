// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker should render correctly. 1`] = `
<Constructor
  backLabel={
    <TmIcon
      name="calendar"
      prefix="tmicon"
      tag="span"
    />
  }
  className=""
  closeOnSelect={true}
  closeOnTab={true}
  columnOfMonth={3}
  dateFormat="YYYY-MM-DD"
  defaultValue=""
  disabled={false}
  entryLabel={
    <TmIcon
      name="clock"
      prefix="tmicon"
      tag="span"
    />
  }
  fullMonthLabel={true}
  input={true}
  inputLabel={
    <TmIcon
      name="calendar"
      prefix="tmicon"
      tag="span"
    />
  }
  inputProps={
    Object {
      "disabled": false,
    }
  }
  isValidDate={[Function]}
  minTimeGap={1}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  renderDay={[Function]}
  strictParsing={true}
  timeConstraints={Object {}}
  timeFormat="HH:mm:ss"
  timeGapUnit="s"
  timeToggleLabelDown={
    <TmIcon
      name="chevron-down"
      prefix="tmicon"
      tag="span"
    />
  }
  timeToggleLabelUp={
    <TmIcon
      name="chevron-up"
      prefix="tmicon"
      tag="span"
    />
  }
  timeViewEntryLabel={null}
  utc={false}
  value="2018-10-11 20:20:20"
  weekDayDisplayMode="short"
/>
`;
